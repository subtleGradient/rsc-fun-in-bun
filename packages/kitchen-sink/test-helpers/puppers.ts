import packageJson from "#package"
import { afterAll, beforeAll } from "bun:test"
import puppeteer, { Browser, ConsoleMessage, Page } from "puppeteer-core"

let browser: Browser

////////////////////////////////////////////////////////////////
// duck punching shennanigans

declare module "puppeteer-core" {
  interface Browser {
    [Symbol.asyncDispose](): Promise<void>
  }
  interface Page {
    [Symbol.asyncDispose](): Promise<void>
    pipeConsoleLogs: boolean
    $pipeConsoleLogs_isEnabled: boolean
  }
}

Browser.prototype[Symbol.asyncDispose] = async function () {
  try {
    if (this.connected) await this.close()
  } catch (e) {
    console.warn("Browser.prototype[Symbol.asyncDispose]", e)
  }
}

Page.prototype[Symbol.asyncDispose] = async function () {
  try {
    if (!this.isClosed()) await this.close({ runBeforeUnload: false })
  } catch (e) {
    console.warn("Page.prototype[Symbol.asyncDispose]", e)
  }
}

Object.defineProperties(Page.prototype, {
  pipeConsoleLogs_isEnabled: { configurable: true, enumerable: false, writable: true, value: false },
  pipeConsoleLogs: {
    configurable: true,
    enumerable: true,
    get(this: Page) { return this.$pipeConsoleLogs_isEnabled }, // prettier-ignore
    set(this: Page, isEnabled: boolean) {
      if ((this.$pipeConsoleLogs_isEnabled = isEnabled)) {
        this.on("console", async (msg: ConsoleMessage) => {
          const args = await Promise.all(msg.args().map(it => it.toString()))
          // @ts-ignore -- trust me bro ðŸ˜Ž
          console[msg.type()](...args)
        })
        this.on("error", console.error)
      } else {
        this.off("console")
        this.off("error")
      }
    },
  },
})

////////////////////////////////////////////////////////////////

beforeAll(async () => {
  const headless = JSON.parse(process.env.PUPPETEER_HEADLESS ?? "false") as boolean
  if (!headless) console.warn("set PUPPETEER_HEADLESS=true to run headless")

  const executablePath =
    process.env.PUPPETEER_EXECUTABLE_PATH ?? "/Applications/Google Chrome.app/Contents/MacOS/Google Chrome"

  console.assert(
    await Bun.file(executablePath).exists(),
    "Chrome executable not found, set PUPPETEER_EXECUTABLE_PATH env var",
  )

  browser = await puppeteer.launch({
    executablePath,
    headless,
    defaultViewport: null,
    waitForInitialPage: false,
    dumpio: true,
    args: [
      "--no-startup-window", // each new page will open in a new window instead of tabs
      `--user-data-dir=${process.env.TMPDIR ?? "/tmp/"}/puppers-${packageJson.name.replace(/[^a-z0-9_.$]/gi, "-")}`,
    ],
  })

  process.once("beforeExit", async () => await browser.close())
})

afterAll(() => browser.close())

export { browser, puppeteer }
