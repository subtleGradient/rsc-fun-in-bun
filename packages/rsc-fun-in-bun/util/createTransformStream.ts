export function createTransformStream(
  transformer: (chunk: Uint8Array) => Uint8Array,
): TransformStream<Uint8Array, Uint8Array> {
  return new TransformStream<Uint8Array, Uint8Array>({
    transform(chunk, controller) {
      // Apply the transformation function to each chunk
      const transformedChunk = transformer(chunk)
      // Enqueue the transformed chunk into the readable side of the stream
      controller.enqueue(transformedChunk)
    },
  })
}

export function createTextTransformStream(transformer: (chunk: string) => string): TransformStream<Uint8Array, Uint8Array> {
  const textDecoder = new TextDecoder()
  const textEncoder = new TextEncoder()

  return new TransformStream<Uint8Array, Uint8Array>({
    transform(chunk, controller) {
      // Decode the chunk to a string
      const textChunk = textDecoder.decode(chunk)
      // Apply the transformation function to the text chunk
      const transformedTextChunk = transformer(textChunk)
      // Encode the transformed text chunk back to Uint8Array and enqueue it
      controller.enqueue(textEncoder.encode(transformedTextChunk))
    },
  })
}
