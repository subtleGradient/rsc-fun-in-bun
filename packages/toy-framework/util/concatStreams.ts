export function concatStreams<T>(...streams: ReadableStream<T>[]): ReadableStream<T> {
  return new ReadableStream<T>({
    async start(ccc) {
      using controller = Object.assign(ccc, { [Symbol.dispose]() { controller.close() } })
      for (const stream of streams) {
        using reader = Object.assign(stream.getReader(), { [Symbol.dispose]() { reader.releaseLock() } })

        while (true) {
          const { done, value } = await reader.read()
          if (done) break
          controller.enqueue(value)
        }
      }
    },
  })
}
