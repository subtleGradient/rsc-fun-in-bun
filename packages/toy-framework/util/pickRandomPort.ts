/** returns a random number between 3000 and 7000 that is not already in use */
export async function pickRandomPort(min = 3000, max = 7000): Promise<number> {
  const port = Math.floor(Math.random() * (max - min)) + min
  const isPortAvailable = await isServerPortAvailable(port)
  return isPortAvailable ? port : await pickRandomPort()
}
function isServerPortAvailable(port: number): Promise<boolean> {
  return new Promise(resolve => {
    try {
      const server = Bun.serve({ port, fetch: () => new Response("Hello!") })
      server.stop(true)
      resolve(true) // available
    } catch (error) {
      resolve(false) // not available
    }
  })
}
