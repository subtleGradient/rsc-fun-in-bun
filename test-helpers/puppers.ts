import { afterAll, beforeAll } from "bun:test"
import puppeteer, { Browser } from "puppeteer-core"

let browser: Browser

beforeAll(async () => {
  const headless = JSON.parse(process.env.PUPPETEER_HEADLESS ?? "false") as boolean
  if (!headless) console.warn("set PUPPETEER_HEADLESS=true to run headless")

  const executablePath =
    process.env.PUPPETEER_EXECUTABLE_PATH ?? "/Applications/Google Chrome.app/Contents/MacOS/Google Chrome"

  console.assert(
    await Bun.file(executablePath).exists(),
    "Chrome executable not found, set PUPPETEER_EXECUTABLE_PATH env var",
  )

  browser = await puppeteer.launch({
    executablePath,
    headless,
    defaultViewport: null,
    waitForInitialPage: false,
    args: [
      "--no-startup-window", // each new page will open in a new window instead of tabs
    ],
  })
})

afterAll(() => browser.close())

// return Object.assign(browser, { [Symbol.dispose]: () => browser.close() })

export { browser, puppeteer }
