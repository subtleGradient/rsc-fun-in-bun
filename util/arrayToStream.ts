export function arrayToStream(...strings: string[]): ReadableStream<Uint8Array> {
  return new ReadableStream<Uint8Array>({
    start(controller) {
      strings.forEach(str => {
        // Convert each string to a Uint8Array and enqueue it
        controller.enqueue(new TextEncoder().encode(str))
      })
      controller.close() // Close the stream after enqueuing all strings
    },
  })
}
