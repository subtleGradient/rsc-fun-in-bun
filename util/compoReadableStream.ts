export function concatStreams<T>(...streams: ReadableStream<T>[]): ReadableStream<T> {
  return new ReadableStream<T>({
    async start(controller) {
      for (const stream of streams) {
        const reader = stream.getReader()

        while (true) {
          const { done, value } = await reader.read()
          if (done) break
          controller.enqueue(value)
        }

        // Close the current reader to ensure it's no longer in use before moving to the next stream
        reader.releaseLock()
      }
      controller.close()
    },
  })
}
