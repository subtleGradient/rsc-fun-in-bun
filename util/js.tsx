export const ts = js
export const tsx = js
export const jsx = js
export function js(strings: TemplateStringsArray, ...values: unknown[]) {
  let source = ""
  for (let i = 0; i < strings.length; i++) {
    source += strings[i]
    if (i < values.length) {
      const value = values[i]
      source += typeof value === "function" ? value.toString() : typeof value === "string" ? value : JSON.stringify(value)
    }
  }
  return source
}

export function html(strings: TemplateStringsArray, ...values: unknown[]) {
  let source = ""
  for (let i = 0; i < strings.length; i++) {
    source += strings[i]
    if (i < values.length) {
      const value = values[i]
      source += typeof value === "function" ? value.toString() : value
    }
  }
  return source
}
