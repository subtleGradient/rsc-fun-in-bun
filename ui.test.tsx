import { type Server } from "bun"
import { afterAll, beforeAll, describe, expect, it } from "bun:test"
import { html, js } from "./examples/js"
import { browser } from "./test-helpers/puppers"
import { arrayToStream } from "./util/arrayToStream"
import { concatStreams } from "./util/compoReadableStream"

if (!1!) {
  describe("the `await using` keyword", async () => {
    it("disposes of the page auto-magically", async () => {
      await using page = Object.assign(await browser.newPage(), { [Symbol.asyncDispose]: () => page.close() })
      await page.goto("about:blank")
      // no need to call page.close() here! 🤯
    })
  })
}

/** this is probably bad for reasons 🤷‍♂️ */
const Port = {
  PORT: 3000,
  get next() {
    return Port.PORT++
  },
}

describe("text server", () => {
  let server: Server
  beforeAll(() => {
    port: Port.next,
      (server = Bun.serve({
        fetch(request) {
          return new Response("Hello from RSC!")
        },
      }))
  })
  afterAll(() => {
    server?.stop()
  })

  describe('GET "/"', () => {
    it("responds with 200", async () => {
      const response = await fetch(server.url.href)
      expect(response.status).toBe(200)
      expect(await response.text()).toBe("Hello from RSC!")
    })
  })
})

describe("HTML server", () => {
  describe("react-dom/server", () => {
    async function htmlStream() {
      const ReactDOMServer = await import("react-dom/server")
      return concatStreams(
        arrayToStream(
          html`<!DOCTYPE html>`,
          `\n<HTML>\n`,
          ReactDOMServer.renderToString(
            <>
              <head>
                <title>Hello from ReactDOMServer.renderToString!</title>
              </head>
            </>,
          ),
          `\n<BODY>\n`,
        ),
        await ReactDOMServer.renderToReadableStream(
          <>
            <h1>Hello from ReactDOMServer.renderToReadableStream!</h1>
            <div id="root" />
            {/* <script type="module" src="/examples/client-bootstrap.js" /> */}
          </>,
        ),
        arrayToStream(
          "\n<!-- ahoy! -->\n",
          html`<script>
            console.log("Hello from RSC!")
          </script>`,
        ),
      )
    }

    let server: Server
    beforeAll(() => {
      server = Bun.serve({
        port: Port.next,
        async fetch(request) {
          return new Response(await htmlStream(), { headers: { "Content-Type": "text/html" } })
        },
      })
    })
    afterAll(() => {
      server?.stop()
    })

    describe('GET "/"', () => {
      it("responds with 200", async () => {
        const response = await fetch(server.url.href)
        expect(response.status).toBe(200)
        expect(await response.text()).toMatchSnapshot()
      })
    })
  })
})

describe.todo("server components", () => { })
describe.todo("async server components", () => { })

describe("client first (classic)", () => {
  async function htmlStream() {
    const ReactDOMServer = await import("react-dom/server")
    return concatStreams(
      arrayToStream(
        html`<!DOCTYPE html><HTML>`,
        ReactDOMServer.renderToString(
          <head>
            <title>Hello from ReactDOMServer.renderToString!</title>
          </head>
        ),
        html`<SCRIPT>${js`
          // fake commonjs environment for react-jsx-dev-runtime
          window.process = { env: { NODE_ENV: "development" } }
          window.require = (module) => {
            if (module === "react") return window.React
            if (module === "react-dom") return window.ReactDOM
            throw new Error("Unknown module: " + module)
          }
          window.exports = window
          `}</SCRIPT>`,
        html`<BODY>`,
      ),
      await ReactDOMServer.renderToReadableStream(
        <>
          <h1>Hello from ReactDOMServer.renderToReadableStream!</h1>
          <div id="root" />
          <script src="/node_modules/react/umd/react.development.js" />
          <script src="/node_modules/react-dom/umd/react-dom.development.js" />
          {/* <script src="/node_modules/react-dom/umd/react-dom-server.browser.development.js" /> */}
          {/* <script src="/node_modules/react-server-dom-webpack/umd/react-server-dom-webpack-server.browser.development.js" /> */}
          {/* <script src="/node_modules/react-server-dom-webpack/umd/react-server-dom-webpack-client.browser.development.js" /> */}
          {/* <script src="https://unpkg.com/babel-standalone@6/babel.js" /> */}
          <script src="/node_modules/react/cjs/react-jsx-dev-runtime.development.js" />
        </>,
      ),
      arrayToStream(
        html`<div id="root">loading...</div>`,
        html`<script>${js`          
          ${function main() {
            const { React, ReactDOM } = window as any as { React: typeof import("react"), ReactDOM: typeof import("react-dom/client") }
            console.log(React.version)

            function ClientComponent() {
              return <div id="generated-by-client">Hello from ClientComponent!</div>
            }

            const root = ReactDOM.hydrateRoot(document.getElementById("root")!, <ClientComponent />)
            // root.render(<ClientComponent />)
          }}
          main()
          `.trim()}</script>`,
      ),
    )
  }

  let server: Server
  beforeAll(() => {
    server = Bun.serve({
      port: Port.next,
      async fetch(request) {
        const url = new URL(request.url)

        if (url.pathname === "/") {
          return new Response(await htmlStream(), { headers: { "Content-Type": "text/html" } })
        }

        const file = Bun.file(__dirname + url.pathname)

        if (await file.exists()) {

          return new Response(file, { headers: { "Content-Type": file.type } })
          // return new Response((await Bun.build({ target: 'browser', entrypoints: [file.name!], format: 'esm' })).outputs[0], { headers: { "Content-Type": file.type } })
        }

        return new Response('404 Not Found', { status: 404 })
      },
    })
  })
  afterAll(() => {
    server?.stop()
  })

  describe("client components", () => {
    describe("render stuff on the client like it's 2014", () => {
      it("renders a component on the client (using fetch)", async () => {
        const response = await fetch(server.url.href)
        expect(response.status).toBe(200)
        // console.log(await response.text())
        expect(await response.text()).toMatchSnapshot()
      })

      it("renders a component on the client (using puppeteer)", async () => {
        await using page = await browser.newPage()
        page.pipeConsoleLogs = true

        await page.goto(server.url.href)
        const version = await page.waitForFunction(() => window?.React?.version).then((x) => x.jsonValue())
        expect(version).toStartWith("19.")
        await page.waitForSelector("#generated-by-client")
      })
    })
  })
})

describe.todo("server actions", () => { })

