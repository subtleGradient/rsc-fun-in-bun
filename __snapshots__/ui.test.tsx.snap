// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`HTML server react-dom/server GET "/" responds with 200 1`] = `
"<!DOCTYPE html>
<HTML>
<head><title>Hello from ReactDOMServer.renderToString!</title></head>
<BODY>
<h1>Hello from ReactDOMServer.renderToReadableStream!</h1><div id="root"></div>
<!-- ahoy! -->
<script>
            console.log("Hello from RSC!")
          </script>"
`;

exports[`client first (classic) client components render stuff on the client like it's 2014 renders a component on the client (using fetch) 1`] = `
"<!DOCTYPE html><HTML><head><title>Hello from ReactDOMServer.renderToString!</title></head><SCRIPT>
          // fake commonjs environment for react-jsx-dev-runtime
          window.process = { env: { NODE_ENV: "development" } }
          window.require = (module) => {
            if (module === "react") return window.React
            if (module === "react-dom") return window.ReactDOM
            throw new Error("Unknown module: " + module)
          }
          window.exports = window
          </SCRIPT><BODY><h1>Hello from ReactDOMServer.renderToReadableStream!</h1><div id="root"></div><script src="/node_modules/react/umd/react.development.js"></script><script src="/node_modules/react-dom/umd/react-dom.development.js"></script><script src="/node_modules/react/cjs/react-jsx-dev-runtime.development.js"></script><div id="root">loading...</div><script>function main() {
      const { React, ReactDOM } = window;
      console.log(React.version);
      function ClientComponent() {
        return jsxDEV("div", {
          id: "generated-by-client",
          children: "Hello from ClientComponent!"
        }, void 0, !1, void 0, this);
      }
      const root = ReactDOM.hydrateRoot(document.getElementById("root"), jsxDEV(ClientComponent, {}, void 0, !1, void 0, this));
    }
          main()</script>"
`;

exports[`client first (modules) client components render stuff on the client like it's 2014 renders a component on the client (using fetch) 1`] = `
"<!DOCTYPE html><HTML><head><title>Hello from ReactDOMServer.renderToString!</title><script type="importmap">{
  "imports": {
    "*": "https://esm.run/",
    "react/jsx-dev-runtime": "/node_modules/react/cjs/react-jsx-dev-runtime.development.js",
    "react/jsx-runtime": "/node_modules/react/cjs/react-jsx-runtime.development.js",
    "react": "/node_modules/react/cjs/react.development.js",
    "react-dom": "https://esm.run/react-dom@19.0.0-canary-a73c3450e-20240329/cjs/react-dom.development.js",
    "react-dom/client": "https://esm.run/react-dom@19.0.0-canary-a73c3450e-20240329/client",
    "react-server-dom-webpack/client": "https://esm.run/react-server-dom-webpack@19.0.0-canary-a73c3450e-20240329/client"
  }
}</script></head><h1>Hello from ReactDOMServer.renderToReadableStream!</h1><div id="root"></div><div id="root">loading...</div><script type=module>let main = function() {
  function ClientComponent() {
    return jsxDEV("div", {
      id: "generated-by-client",
      children: "Hello from ClientComponent!"
    }, undefined, false, undefined, this);
  }
  const root = ReactDOM.hydrateRoot(document.getElementById("root"), jsxDEV(ClientComponent, {}, undefined, false, undefined, this));
};
import JSX from "react/jsx-dev-runtime";
const { jsxDEV, Fragment } = JSX;
import React from "react";
import ReactDOM from "react-dom/client";
console.log(React.version);
console.log(jsxDEV);
Object.assign(window, { React, ReactDOM, jsxDEV, Fragment });
main();
</script>"
`;
